src/test/osd/TestOSDMap.cc

```diff
@@ -1366,6 +1366,94 @@ TEST_F(OSDMapTest, CleanPGUpmaps) {
+    osd_name << "osd." << i;
+    host_name << "host-" << index;
+    move_to.push_back("root=default");
+    string host_loc = "host=" + host_name.str();
+    move_to.push_back(host_loc);
+    int r = crush_move(osdmap, osd_name.str(), move_to);
+    ASSERT_EQ(0, r);
+  }
+
+  // Create a CRUSH rule for pg_upmap_primary
+  const string upmap_primary_rule = "upmap_primary";
+  int upmap_primary_rule_no = crush_rule_create_replicated(
+    upmap_primary_rule, "default", "host");
+  ASSERT_LT(0, upmap_primary_rule_no);
+
+  // Create a replicated pool with an initial PG count
+  OSDMap::Incremental new_pool_inc(osdmap.get_epoch() + 1);
+  new_pool_inc.new_pool_max = osdmap.get_pool_max();
+  new_pool_inc.fsid = osdmap.get_fsid();
+  pg_pool_t empty;
+  uint64_t upmap_primary_pool_id = ++new_pool_inc.new_pool_max;
+  pg_pool_t *p = new_pool_inc.get_new_pool(upmap_primary_pool_id, &empty);
+  p->size = 2;
+  p->set_pg_num(64);
+  p->set_pgp_num(64);
+  p->type = pg_pool_t::TYPE_REPLICATED;
+  p->crush_rule = upmap_primary_rule_no;
+  p->set_flag(pg_pool_t::FLAG_HASHPSPOOL);
+  new_pool_inc.new_pool_names[upmap_primary_pool_id] = "upmap_primary_pool";
+  osdmap.apply_incremental(new_pool_inc);
+
+  // Create pg_upmap_primary mappings for each PG in the pool
+  OSDMap::Incremental pending_inc(osdmap.get_epoch() + 1);
+  for (int pg_num = 0; pg_num < 64; ++pg_num) {
+    pg_t rawpg(pg_num, upmap_primary_pool_id);
+    pg_t pgid = osdmap.raw_pg_to_pg(rawpg);
+    vector<int> up;
+    int up_primary;
+    osdmap.pg_to_raw_up(pgid, &up, &up_primary);
+    int new_primary = (up_primary + 1) % get_num_osds();
+    pending_inc.new_pg_upmap_primary[pgid] = new_primary;
+  }
+  OSDMap nextmap;
+  nextmap.deepish_copy_from(osdmap);
+  nextmap.apply_incremental(pending_inc);
+  ASSERT_EQ(nextmap.get_num_pg_upmap_primaries(), 64);
+
+  // Reduce the number of PGs, causing some PGs to be removed
+  OSDMap::Incremental remove_pool_inc(nextmap.get_epoch() + 1);
+  remove_pool_inc.old_pools.insert(nextmap.lookup_pg_pool_name("upmap_primary_pool"));
+  nextmap.apply_incremental(remove_pool_inc);
+
+  OSDMap::Incremental reduced_pool_inc(nextmap.get_epoch() + 1);
+  pg_pool_t new_empty;
+  pg_pool_t *reduced_pool = reduced_pool_inc.get_new_pool(upmap_primary_pool_id, &new_empty);
+  reduced_pool->size = 2;
+  reduced_pool->set_pg_num(8);
+  reduced_pool->set_pgp_num(8);
+  reduced_pool->type = pg_pool_t::TYPE_REPLICATED;
+  reduced_pool->crush_rule = upmap_primary_rule_no;
+  reduced_pool->set_flag(pg_pool_t::FLAG_HASHPSPOOL);
+  reduced_pool_inc.new_pool_names[upmap_primary_pool_id] = "upmap_primary_pool";
+  nextmap.apply_incremental(reduced_pool_inc);
+
+  // Clean stale pg_upmap_primary entries
+  OSDMap::Incremental cleanup_inc(nextmap.get_epoch() + 1);
+  clean_pg_upmaps(g_ceph_context, nextmap, cleanup_inc);
+  nextmap.apply_incremental(cleanup_inc);
+
+  // Ensure stale pg_upmap_primary entries are removed
+  ASSERT_EQ(nextmap.get_num_pg_upmap_primaries(), 8);
```

Comment:
This is OK. The only thing is that you create illegal pg_upmap_prims (where the prim is not in the acting set of the pg, which is blocked in the code). The clean command should also remove such records. 
The "normal" path would in line 1423 take the secondary and not (primary + 1)  % num_osds.
In addition I would add a test that creates illegal pg_primaries on purpose (where the primary is not in the PG) and see that the clean removes them (this can be done without changing the pg_num in te test process).